src/transformStateWithClones.js

```diff
@@ -7,7 +7,27 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const result = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const keys of action.keysToRemove) {
+          delete stateCopy[keys];
+        }
+        break;
 
+      default:
+        stateCopy = {};
+    }
+    result.push({ ...stateCopy });
```

Comment:
[DONT REPEAT YOURSELF] - It is better to push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after the current action is processed. This is already done correctly in your code.